src/rfq/protocols/bebop/client.rs

```diff
@@ -0,0 +1,395 @@
+                .body(())
+                .expect("Failed to build request");
+
+            // Connect to Bebop WebSocket with custom headers
+            let (ws_stream, _) = match connect_async_with_config(request, None, false).await {
+                Ok(connection) => connection,
+                Err(e) => {
+                    tracing::error!("Failed to connect to Bebop WebSocket: {}", e);
+                    return;
+                }
+            };
+
+            let (_, mut ws_receiver) = ws_stream.split();
+
+            // TODO: What to do if the program crashes and restarts? Should we be retrieving these components
+            // from the database?
+            let mut current_components: HashMap<String, ComponentWithState> = HashMap::new();
+
+            while let Some(msg) = ws_receiver.next().await {
+                match msg {
+                    Ok(Message::Text(text)) => {
+                        if let Ok(price_data_map) = serde_json::from_str::<BebopPriceMessage>(&text) {
+                            for (pair, price_data) in price_data_map {
+                                if pairs.contains(&pair) {
+
+                                    let tvl = client.calculate_tvl(&price_data);
+                                    if tvl < tvl_threshold {
+                                        continue;
+                                    }
+
+                                    let component_id = format!("bebop_{}", pair.replace("/", "_"));
+                                    let component_with_state = client.create_component_with_state(&pair, &price_data);
+
+                                    let new_components = HashMap::from([(component_id.clone(), component_with_state)]);
+
+                                    let removed_components: Vec<String> = current_components.keys()
+                                        .filter(|id| !new_components.contains_key(*id))
+                                        .cloned()
+                                        .collect();
+
+                                    current_components = new_components.clone();
+
+                                    let snapshot = Snapshot {
+                                        states: new_components,
+                                        vm_storage: HashMap::new(),
+                                    };
+
+                                    let timestamp = price_data.last_update_ts as u64;
+                                    let msg = StateSyncMessage::<TimestampHeader> {
+                                        header: TimestampHeader { timestamp },
+                                        snapshots: snapshot,
+                                        deltas: None,
+                                        removed_components: removed_components.into_iter().map(|id| (id, Default::default())).collect(),
+                                    };
+
+                                    yield (component_id, msg);
```

Comment:
wait.. you are emitting a message per pair right? 🤔 We wanted a new message with all pairs info (removed included)

Reply:
ah true!